rs/execution_environment/src/execution_environment.rs

```diff
@@ -1622,9 +1623,30 @@ impl ExecutionEnvironment {
                 )
             }
             CanisterMessageOrTask::Message(CanisterMessage::Request(request)) => {
+                if let Some(metadata) = &request.metadata {
+                    let message_type_label = "request";
+                    let call_tree_depth_label = &metadata.call_tree_depth().to_string();
+                    let latency = current_time()
+                        .saturating_duration_since(*metadata.call_tree_start_time())
+                        .as_secs_f64();
+                    self.metrics
+                        .canister_message_queue_latency
+                        .with_label_values(&[message_type_label, call_tree_depth_label])
+                        .observe(latency);
+                }
                 CanisterCall::Request(request)
             }
             CanisterMessageOrTask::Message(CanisterMessage::Ingress(ingress)) => {
+                let message_type_label = "ingress";
+                let call_tree_depth_label = "0"; // Ingress messages do not have a call tree.
+                let now = current_time();
+                let expiry_duration = expiry_time_from_now().saturating_duration_since(now);
+                let start = ingress.expiry_time.saturating_sub(expiry_duration);
+                let latency = now.saturating_duration_since(start).as_secs_f64();
```

Comment:
Based on your PR description you are very likely aware of this, but I will still mention that the expiry time can be set by the caller to an arbitrary value, i.e. much lower than the maximum 5 minutes. So this calculation may also be way off.

(In fact, I seem to remember subnets where this is consistently significantly less than 5 minutes. There's an `mr_unreliable_induct_ingress_message_duration_seconds` metric that uses similar logic to estimate ingress induction latency. You can go [here](https://grafana.mainnet.dfinity.network/d/ic-health/ic-health?orgId=1&editPanel=43), unhide the first query and try selecting various subnets.)